File: Neetcode/Array&Hashing/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        Most common solutions for string problems are done using
        Set() or Dict().
        Use dict to record the mapping of
            (Difference of A[i] and target): i
        When another value A[j] matches a key in the dict, return.
        """
        
        valToIndex = {}
        for i in range(len(nums)):
            if nums[i] in valToIndex:
                return [i, valToIndex[nums[i]]]
            else:
                delta = target - nums[i]
                valToIndex[delta] = i
        
# @lc code=end

File: Java/1.two-sum.java
/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start

import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> valToIndex = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int theOther = target - nums[i];
            if (valToIndex.containsKey(theOther)) {
                return new int[] { valToIndex.get(theOther), i };
            }
            valToIndex.put(nums[i], i);
        }
        return new int[0];
    }
}
// @lc code=end
